design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,136 @@
+
+### Non-Goals
+- Define an upper limit to the number of network devices that can be defined. 
+
+## Proposal
+Users are to express the potential amount of network devices through
+the VMI spec.
+Internally, kubevirt will use libvirt logic to generate the needed
+devices to support the amount specified (e.g. PCIe controllers).
+
+Using hotplug, users may add network devices up to the amount specified
+by the new user *network-interface-list-capacity* input.
+
+### Definition of Personas
+- VM/VMI administrator: Any user who can create VM and VMI objects.
+
+### User Stories
+- TBD
+
+### API Extensions
+The VMI Spec will be extended to include a new field, named `networkInterfaceListCapacity`.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 1 and 32.
+- In case the value is less than the amount of interfaces defined already,
+  it has no effect.
+
+#### VMI Spec example
+```yaml
+
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-iface-capacity-example
+spec:
+  domain:
+    devices:
+      disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      interfaces:
+        - masquerade: {}
+          name: default
+        - bridge: {}
+          name: secondary
+      networkInterfaceListCapacity: 5
+      rng: {}
+  networks:
+    - name: default
+      pod: {}
+    - multus:
+        networkName: some-nad
+      name: secondary 
+  volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+      name: containerdisk
+```
+
+## Design Details
+The solution is taking advantage of the existing libvirt capability to auto-generate
+the needed configuration (e.g. controllers) when specifying network interfaces.
+
+The main change is in the logic of creating a domain in the virt-launcher.
+
+This is the main flow suggested:
+- Base on the network-interface-list-capacity and the current defined interfaces, generate
+  [interface-placeholders](#network-interface-placeholder) and add them to the domain spec.
+- Following the creation of the domain configuration, remove these placeholders
+  from he configuration and start the domain without them.
+- At this stage, the domain has been started without the additional interface devices
+  but with the additional configuration (e.g. controllers) they originally generated.
+
+#### Network Interface Placeholder
+The interface placeholder is a temporary dummy definition of an interface in the domain spec
+of libvirt. It is used to trigger libvirt to auto-generate dependent devices (e.g. PCIe controllers).
+
+After the domain configuration is constructed, these placeholders are removed.
+
+This is an example of such a placeholder:
+```xml
+<interface type='ethernet'>
+  <target dev='random123' managed='no'/>
+  <model type='virtio-non-transitional'/>
+  <mtu size='1500'/>
+  <alias name='k6t-placeholder-1'/>
```

Comment:
The alias is what makes the whole thing work :)

The idea is that a YAML like

```yaml
devices:
  interfaces:
    # repeated cur times
    - name: ...
  networkInterfaceListCapacity: cap # with cap >= cur
```

will result in the XML

```xml
<!-- repeated cur times -->
<interface>
  ...
</interface>

<!-- repeated (cap - cur) times -->
<interface>
  <alias name='k6t-placeholder-...'/>
</interface>
```

This will result in libvirt creating a PCI topology that can accomodate up to `cap` network interfaces.

Before the VM is started, all the `cap - cur` placeholder interfaces will be dropped, leaving the PCI controllers necessary to get back to `cap` network interfaces behind.

Marking the placeholder interfaces with unique aliases is needed in order to be able to easily and programmatically remove them.

Reply:
What target.dev field is not enough for this dropping logic? After the domain is created, cannot we remove all the interfaces with target.dev with "random" prefix.

Reply:
Yeah, that could work too. It's really not important how placeholders are marked as such, as long as we can clearly tell them apart from real devices.

Reply:
Fair point, I will remove it. This is also a very low detail which we do not need to commit to in a design anyway.